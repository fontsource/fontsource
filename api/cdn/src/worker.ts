import { error } from 'itty-router';

import router, { corsify } from './router';

export default {
	async fetch(
		request: Request,
		env: Env,
		ctx: ExecutionContext,
	): Promise<Response> {
		return await router
			.handle(request, env, ctx)
			.catch((error_) => {
				console.error(error_);
				const resp = error(error_);
				// Our jsDelivr proxy will fallback to one hour if no cache-control is set
				resp.headers.set('Cache-Control', 'max-age=0');
				return resp;
			})
			.then(corsify);
	},
};
