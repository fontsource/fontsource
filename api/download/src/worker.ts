import router from './router';
import { error } from 'itty-router';

export default {
	async fetch(
		request: Request,
		env: Env,
		ctx: ExecutionContext
	): Promise<Response> {
		return router.handle(request, env, ctx).catch((err) => {
			console.error(err);
			return error(err);
		});
	},
};
