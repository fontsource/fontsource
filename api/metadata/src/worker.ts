import fontsRouter from './fonts/router';
import fontlistRouter from './fontlist/router';
import downloadRouter from './download/router';
import { error } from 'itty-router';

export default {
	async fetch(
		request: Request,
		env: Env,
		ctx: ExecutionContext
	): Promise<Response> {
		try {
			const url = new URL(request.url);

			if (url.pathname.startsWith('/fontlist')) {
				return fontlistRouter.handle(request, env, ctx).catch((err) => {
					console.error(err);
					return error(err);
				});
			}

			if (url.pathname.startsWith('/v1/fonts')) {
				return fontsRouter.handle(request, env, ctx).catch((err) => {
					console.error(err);
					return error(err);
				});
			}

			if (url.pathname.startsWith('/v1/download')) {
				return downloadRouter.handle(request, env, ctx).catch((err) => {
					console.error(err);
					return error(err);
				});
			}

			return error(
				404,
				'Not Found. Please refer to the Fontsource API documentation: https://fontsource.org/docs/api'
			);
		} catch (e) {
			console.error(e);
			return error(500, 'Internal Server Error.');
		}
	},
};
