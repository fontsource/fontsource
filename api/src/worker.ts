import fontsRouter from './fonts/router';

export default {
	async fetch(
		request: Request,
		env: Env,
		ctx: ExecutionContext
	): Promise<Response> {
		const url = new URL(request.url);

		if (url.pathname.startsWith('/v1/fonts/')) {
			return fontsRouter.handle(request);
		}

		return new Response('Not Found.', { status: 404 });
	},
};
