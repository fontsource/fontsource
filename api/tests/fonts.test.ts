import { beforeEach, expect, it } from 'vitest';
const describe = setupMiniflareIsolatedStorage();
import worker from '../src/worker';
import { mockMetadata } from './helpers';

describe('fonts worker', () => {
	const env = getMiniflareBindings() as Env;
	const ctx = new ExecutionContext();

	beforeEach(async () => {
		mockMetadata();
	});

	it('should return array of metadata default request', async () => {
		const request = new Request('http://localhost:8787/v1/fonts');
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(200);

		const body = await response.json();
		expect(body).toMatchSnapshot();
	});

	it('should filter a single query param (string)', async () => {
		const request = new Request('http://localhost:8787/v1/fonts?type=other');
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(200);

		const body = await response.json();
		expect(body).toMatchSnapshot();
	});

	it('should filter a single query param (string array)', async () => {
		const request = new Request(
			'http://localhost:8787/v1/fonts?subsets=arabic'
		);
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(200);

		const body = await response.json();
		expect(body).toMatchSnapshot();
	});

	it('should filter a single query param (number array)', async () => {
		const request = new Request('http://localhost:8787/v1/fonts?weights=900');
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(200);

		const body = await response.json();
		expect(body).toMatchSnapshot();
	});

	it('should filter a single query param (boolean)', async () => {
		const request = new Request('http://localhost:8787/v1/fonts?variable=true');
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(200);

		const body = await response.json();
		expect(body).toMatchSnapshot();
	});

	it('should filter multiple query params', async () => {
		const request = new Request(
			'http://localhost:8787/v1/fonts?category=display&variable=false'
		);
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(200);

		const body = await response.json();
		expect(body).toMatchSnapshot();
	});

	it('should filter comma separated query params', async () => {
		const request = new Request(
			'http://localhost:8787/v1/fonts?category=display,handwriting'
		);
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(200);

		const body = await response.json();
		expect(body).toMatchSnapshot();
	});

	it('should return 400 for invalid query string', async () => {
		const request = new Request(`http://localhost:8787/v1/fonts?invalid`);
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(400);
	});
});
