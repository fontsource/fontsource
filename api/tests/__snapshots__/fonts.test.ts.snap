// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fonts worker > should filter a single query param (array) 1`] = `[]`;

exports[`fonts worker > should filter a single query param (boolean) 1`] = `
[
  {
    "category": "sans-serif",
    "defSubset": "latin",
    "family": "Advent Pro",
    "id": "advent-pro",
    "lastModified": "2023-03-21",
    "license": "OFL-1.1",
    "styles": [
      "italic",
      "normal",
    ],
    "subsets": [
      "cyrillic",
      "cyrillic-ext",
      "greek",
      "latin",
      "latin-ext",
    ],
    "type": "google",
    "variable": true,
    "weights": [
      100,
      200,
      300,
      400,
      500,
      600,
      700,
      800,
      900,
    ],
  },
  {
    "category": "icons",
    "defSubset": "latin",
    "family": "Material Symbols Outlined",
    "id": "material-symbols-outlined",
    "lastModified": "2023-04-27",
    "license": "Apache-2.0",
    "styles": [
      "normal",
    ],
    "subsets": [
      "latin",
    ],
    "type": "google",
    "variable": true,
    "weights": [
      100,
      200,
      300,
      400,
      500,
      600,
      700,
    ],
  },
]
`;

exports[`fonts worker > should filter a single query param (number array) 1`] = `
[
  {
    "category": "sans-serif",
    "defSubset": "latin",
    "family": "Advent Pro",
    "id": "advent-pro",
    "lastModified": "2023-03-21",
    "license": "OFL-1.1",
    "styles": [
      "italic",
      "normal",
    ],
    "subsets": [
      "cyrillic",
      "cyrillic-ext",
      "greek",
      "latin",
      "latin-ext",
    ],
    "type": "google",
    "variable": true,
    "weights": [
      100,
      200,
      300,
      400,
      500,
      600,
      700,
      800,
      900,
    ],
  },
  {
    "category": "other",
    "defSubset": "latin",
    "family": "YakuHanJP",
    "id": "yakuhanjp",
    "lastModified": "2023-05-15",
    "license": "OFL-1.1",
    "styles": [
      "normal",
    ],
    "subsets": [
      "latin",
    ],
    "type": "other",
    "variable": false,
    "weights": [
      100,
      200,
      300,
      400,
      500,
      700,
      900,
    ],
  },
]
`;

exports[`fonts worker > should filter a single query param (string array) 1`] = `[]`;

exports[`fonts worker > should filter a single query param (string) 1`] = `
[
  {
    "category": "other",
    "defSubset": "latin",
    "family": "YakuHanJP",
    "id": "yakuhanjp",
    "lastModified": "2023-05-15",
    "license": "OFL-1.1",
    "styles": [
      "normal",
    ],
    "subsets": [
      "latin",
    ],
    "type": "other",
    "variable": false,
    "weights": [
      100,
      200,
      300,
      400,
      500,
      700,
      900,
    ],
  },
]
`;

exports[`fonts worker > should filter comma separated query params 1`] = `[]`;

exports[`fonts worker > should filter multiple query params 1`] = `[]`;

exports[`fonts worker > should return array of metadata default request 1`] = `
[
  {
    "category": "sans-serif",
    "defSubset": "latin",
    "family": "Abel",
    "id": "abel",
    "lastModified": "2022-09-22",
    "license": "OFL-1.1",
    "styles": [
      "normal",
    ],
    "subsets": [
      "latin",
    ],
    "type": "google",
    "variable": false,
    "weights": [
      400,
    ],
  },
  {
    "category": "sans-serif",
    "defSubset": "latin",
    "family": "Advent Pro",
    "id": "advent-pro",
    "lastModified": "2023-03-21",
    "license": "OFL-1.1",
    "styles": [
      "italic",
      "normal",
    ],
    "subsets": [
      "cyrillic",
      "cyrillic-ext",
      "greek",
      "latin",
      "latin-ext",
    ],
    "type": "google",
    "variable": true,
    "weights": [
      100,
      200,
      300,
      400,
      500,
      600,
      700,
      800,
      900,
    ],
  },
  {
    "category": "icons",
    "defSubset": "latin",
    "family": "Material Icons",
    "id": "material-icons",
    "lastModified": "2023-03-08",
    "license": "Apache-2.0",
    "styles": [
      "normal",
    ],
    "subsets": [
      "latin",
    ],
    "type": "google",
    "variable": false,
    "weights": [
      400,
    ],
  },
  {
    "category": "icons",
    "defSubset": "latin",
    "family": "Material Symbols Outlined",
    "id": "material-symbols-outlined",
    "lastModified": "2023-04-27",
    "license": "Apache-2.0",
    "styles": [
      "normal",
    ],
    "subsets": [
      "latin",
    ],
    "type": "google",
    "variable": true,
    "weights": [
      100,
      200,
      300,
      400,
      500,
      600,
      700,
    ],
  },
  {
    "category": "other",
    "defSubset": "latin",
    "family": "YakuHanJP",
    "id": "yakuhanjp",
    "lastModified": "2023-05-15",
    "license": "OFL-1.1",
    "styles": [
      "normal",
    ],
    "subsets": [
      "latin",
    ],
    "type": "other",
    "variable": false,
    "weights": [
      100,
      200,
      300,
      400,
      500,
      700,
      900,
    ],
  },
]
`;
